En esta fase entrenamos y evaluamos los modelos supervisados del proyecto IA-CANCER-LOGISTIC-RF, utilizando los datos preparados en la Fase 2. A partir de estos insumos, se construyen dos clasificadores: Regresión Logística, que trabaja con las variables estandarizadas, y Random Forest, que utiliza las características originales.

El notebook desarrolla el flujo completo de entrenamiento: carga de los conjuntos de train y test, creación y ajuste de ambos modelos, cálculo de métricas como accuracy, precisión, recall y F1-score, y análisis de sus matrices de confusión para entender los aciertos y errores de cada algoritmo. Finalmente, los modelos entrenados se guardan en la carpeta models/ para su uso en etapas posteriores.

Con esto se completa el ciclo de entrenamiento y los modelos quedan listos para integrarse en un flujo de predicción o una aplicación externa.

In [15]:
# Paso 1: Importar librerías para modelos supervisados y métricas

import numpy as np
import pandas as pd

# Modelos
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

# Métricas de evaluación (Accuracy, Precision, Recall, F1, matriz de confusión)
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report
)

# Guardar modelos entrenados
import joblib
import os

print("✅ Librerías para modelos y métricas importadas correctamente.")

✅ Librerías para modelos y métricas importadas correctamente.


In [17]:
# Cargar conjuntos procesados desde disco
X_train_logreg = joblib.load("data/processed/X_train_logreg.pkl")
X_test_logreg  = joblib.load("data/processed/X_test_logreg.pkl")

X_train_rf = joblib.load("data/processed/X_train_rf.pkl")
X_test_rf  = joblib.load("data/processed/X_test_rf.pkl")

y_train_final = joblib.load("data/processed/y_train_final.pkl")
y_test_final  = joblib.load("data/processed/y_test_final.pkl")

print("✅ Conjuntos cargados correctamente en Fase 3.")


✅ Conjuntos cargados correctamente en Fase 3.


# Modelo de Regresión Logística

##  Instanciar el modelo de Regresión Logística
A continuación, vamos a instanciar el modelo de Regresión Logística. En este punto únicamente creamos el modelo; el entrenamiento se realizará más adelante. Para ello, agrega una nueva celda en tu notebook y escribe:

In [ ]:
# Instanciar el modelo de Regresión Logística

model_logreg = LogisticRegression(
    max_iter=1000,       # más iteraciones para asegurar convergencia
    solver='lbfgs',      # funciona bien con datasets medianos
    random_state=42
)

print("Modelo de Regresión Logística instanciado.")


Modelo de Regresión Logística instanciado.


## Entrenar el modelo de Regresión Logística


In [ ]:
# Entrenar el modelo de Regresión Logística
model_logreg.fit(X_train_logreg, y_train_final)
print("Modelo de Regresión Logística entrenado correctamente.")

Modelo de Regresión Logística entrenado correctamente.


## Evaluación del modelo

#### Realizar predicciones con el modelo entrenado
Antes de calcular las métricas, necesitamos obtener las predicciones de la Regresión Logística sobre el conjunto de prueba (X_test_logreg). Estas predicciones devolverán una lista de valores 0 y 1, donde 0 corresponde a tumores benignos y 1 a tumores malignos.

In [ ]:
# Realizar predicciones con el modelo de Regresión Logística

y_pred_logreg = model_logreg.predict(X_test_logreg)

print("Predicciones generadas para el conjunto de prueba.")


Predicciones generadas para el conjunto de prueba.


#### Calcular métricas de evaluación
Ahora que ya tenemos las predicciones del modelo, el siguiente paso es calcular las métricas de evaluación para entender qué tan bien está funcionando la Regresión Logística. Estas métricas te permitirán interpretar el desempeño del modelo: accuracy, precisión, recall y F1-score, cada una enfocada en un aspecto diferente de la clasificación.

In [ ]:
# Paso 5: Evaluación del modelo de Regresión Logística

accuracy_logreg  = accuracy_score(y_test_final, y_pred_logreg)
precision_logreg = precision_score(y_test_final, y_pred_logreg)
recall_logreg    = recall_score(y_test_final, y_pred_logreg)
f1_logreg        = f1_score(y_test_final, y_pred_logreg)

print("Resultados del Modelo de Regresión Logística:")
print(f"Accuracy : {accuracy_logreg:.4f}")
print(f"Precisión: {precision_logreg:.4f}")
print(f"Recall   : {recall_logreg:.4f}")
print(f"F1-score : {f1_logreg:.4f}")


📊 Resultados del Modelo de Regresión Logística:
Accuracy : 0.9649
Precisión: 0.9750
Recall   : 0.9286
F1-score : 0.9512


Los resultados obtenidos muestran que el modelo de Regresión Logística presenta un desempeño muy sólido. La accuracy alcanzada es de 0.9649, lo que significa que el modelo acierta aproximadamente el 96.5% de las predicciones. La precisión es de 0.9750, indicando que cuando el modelo predice un caso como “maligno”, casi siempre corresponde a un tumor maligno real. El recall es de 0.9286, por lo que logra identificar la mayoría de los tumores malignos presentes en el conjunto de prueba. Finalmente, el F1-score de 0.9512 confirma un buen equilibrio entre precisión y recall.

Como siguiente paso, es importante observar no solo los valores agregados de las métricas, sino también cómo se distribuyen los aciertos y errores del modelo. Para ello, generaremos la matriz de confusión, que nos permitirá visualizar con mayor detalle qué tipos de predicciones acierta y en cuáles se equivoca.

In [ ]:
#  Matriz de confusión del modelo de Regresión Logística

cm_logreg = confusion_matrix(y_test_final, y_pred_logreg)

print(" Matriz de confusión (Regresión Logística):")
print(cm_logreg)


🔍 Matriz de confusión (Regresión Logística):
[[71  1]
 [ 3 39]]


La matriz de confusión confirma el buen rendimiento del modelo. La Regresión Logística clasificó correctamente **71 de 72 tumores benignos** y **39 de 42 malignos**, mostrando una buena capacidad para diferenciar ambas clases. Los errores fueron pocos: un caso benigno fue marcado como maligno (falso positivo) y tres casos malignos se clasificaron como benignos (falsos negativos).

Aunque los falsos negativos son especialmente sensibles en contextos médicos, su número es bajo en comparación con el total evaluado, lo que indica que el modelo logra una detección adecuada de los casos críticos. En conjunto, estos resultados refuerzan que la Regresión Logística está funcionando de manera estable y efectiva para este problema.

# Modelo de Random Forest

## Instanciar el modelo de Random Forest

In [27]:
# Instanciar el modelo de Random Forest

model_rf = RandomForestClassifier(
    n_estimators=200,   # número de árboles
    max_depth=None,     # profundidad libre
    random_state=42
)

print(" Modelo de Random Forest instanciado correctamente.")


 Modelo de Random Forest instanciado correctamente.


## Entrenar el modelo de Random Forest

In [28]:
# Entrenar el modelo de Random Forest

model_rf.fit(X_train_rf, y_train_final)

print("Modelo Random Forest entrenado correctamente.")


Modelo Random Forest entrenado correctamente.


## Evaluación del modelo Ranfom Forest

### Realizar predicciones con Random Forest

Ahora vamos a generar las predicciones del modelo Random Forest utilizando el conjunto de prueba sin escalar (X_test_rf). Este paso nos permitirá evaluar su desempeño y compararlo más adelante con el de la Regresión Logística. Para ello, crea una nueva celda y ejecuta el siguiente código:

In [29]:
# Paso 8: Predicciones con el modelo Random Forest

y_pred_rf = model_rf.predict(X_test_rf)

print("Predicciones generadas para Random Forest.")


Predicciones generadas para Random Forest.


#### Calcular métricas de evaluación
Ahora que ya generamos las predicciones del Random Forest, vamos a evaluar su desempeño calculando accuracy, precisión, recall y F1-score, utilizando las mismas variables de referencia del modelo anterior (y_test_final y y_pred_rf). Esto nos permitirá comparar ambos modelos bajo las mismas condiciones.

In [30]:
# Evaluación del modelo Random Forest

accuracy_rf  = accuracy_score(y_test_final, y_pred_rf)
precision_rf = precision_score(y_test_final, y_pred_rf)
recall_rf    = recall_score(y_test_final, y_pred_rf)
f1_rf        = f1_score(y_test_final, y_pred_rf)

print("📊 Resultados del Modelo Random Forest:")
print(f"Accuracy : {accuracy_rf:.4f}")
print(f"Precisión: {precision_rf:.4f}")
print(f"Recall   : {recall_rf:.4f}")
print(f"F1-score : {f1_rf:.4f}")


📊 Resultados del Modelo Random Forest:
Accuracy : 0.9649
Precisión: 1.0000
Recall   : 0.9048
F1-score : 0.9500


In [31]:
# Paso 10: Matriz de confusión del modelo Random Forest

cm_rf = confusion_matrix(y_test_final, y_pred_rf)

print("Matriz de confusión (Random Forest):")
print(cm_rf)

Matriz de confusión (Random Forest):
[[72  0]
 [ 4 38]]


El modelo Random Forest obtuvo un desempeño general muy sólido, con un accuracy de 0.9649 y un F1-score de 0.9500, valores prácticamente equivalentes a los de la Regresión Logística. Sin embargo, la distribución de errores muestra un comportamiento diferente en la forma como el modelo clasifica cada tipo de tumor.

La matriz de confusión evidencia que el modelo clasificó correctamente 72 de 72 tumores benignos, sin cometer ningún falso positivo, lo cual explica su precisión perfecta (1.0000) al predecir tumores malignos. Este comportamiento indica que Random Forest es un modelo altamente conservador: solo predice “maligno” cuando está muy seguro.

Por otra parte, el modelo presentó 4 falsos negativos, es decir, casos malignos que fueron clasificados como benignos. Esto se refleja en un recall de 0.9048, ligeramente inferior al obtenido con la Regresión Logística. En aplicaciones médicas, este tipo de error es especialmente importante porque implica no detectar casos críticos.

En conjunto, Random Forest ofrece una clasificación muy precisa para tumores benignos y un buen rendimiento general, aunque con una mayor tendencia a pasar por alto algunos casos malignos debido a su carácter conservador.

# Comparación entre Regresión Logística y Random Forest
Ahora vamos a organizar los resultados de ambos modelos en una pequeña tabla de comparación.

In [33]:
# Comparación entre modelos sin usar

resultados = pd.DataFrame({
    "Métrica": ["Accuracy", "Precisión", "Recall", "F1-score"],
    "Regresión Logística": [
        accuracy_logreg,
        precision_logreg,
        recall_logreg,
        f1_logreg
    ],
    "Random Forest": [
        accuracy_rf,
        precision_rf,
        recall_rf,
        f1_rf
    ]
})

print("📊 Comparación entre modelos:")
print(resultados.to_string(index=False, formatters={"Regresión Logística": "{:.4f}".format,
                                                   "Random Forest": "{:.4f}".format}))


📊 Comparación entre modelos:
  Métrica Regresión Logística Random Forest
 Accuracy              0.9649        0.9649
Precisión              0.9750        1.0000
   Recall              0.9286        0.9048
 F1-score              0.9512        0.9500


Ambos modelos presentan un desempeño muy similar en términos de exactitud general (accuracy ≈ 0.9649), lo cual indica que los dos clasifican correctamente la gran mayoría de los casos del conjunto de prueba. Sin embargo, existen diferencias importantes en la forma como cada modelo comete errores.

La Regresión Logística ofrece un mejor equilibrio entre precisión y recall, con un F1-score ligeramente superior (0.9512). Su recall (0.9286) es más alto que el de Random Forest, lo que implica que detecta un mayor número de tumores malignos, reduciendo la cantidad de falsos negativos. Este comportamiento es particularmente valioso en contextos médicos, donde es preferible identificar la mayor cantidad posible de casos malignos.

Por otro lado, Random Forest se caracteriza por tener una precisión perfecta (1.0000): no genera falsos positivos. Sin embargo, su recall es menor (0.9048), lo que indica que deja pasar algunos tumores malignos al clasificarlos como benignos. Esto refleja un modelo más conservador para predecir malignidad: solo etiqueta un caso como maligno cuando tiene una seguridad muy alta.

En conjunto, ambos modelos son adecuados para el problema, pero la Regresión Logística ofrece un mejor balance general, mientras que Random Forest prioriza evitar falsos positivos a costa de aumentar los falsos negativos.

# Guardar los modelos entrenados

In [43]:
import os
import joblib

# Obtener la ruta absoluta de la carpeta raíz del proyecto
root_dir = os.path.dirname(os.getcwd())   # sube desde /notebooks a la raíz

# Construir la ruta absoluta hacia /models
models_path = os.path.join(root_dir, "models")

# Crear carpeta models si no existe
os.makedirs(models_path, exist_ok=True)

# Guardar modelos usando la ruta absoluta
joblib.dump(model_logreg, os.path.join(models_path, "model_logreg.pkl"))
joblib.dump(model_rf, os.path.join(models_path, "model_rf.pkl"))

print("💾 Modelos guardados correctamente en:", models_path)


💾 Modelos guardados correctamente en: c:\Users\Familia\Downloads\IA\ia-cancer-logistic-rf\models
